File: server_py/abi/providers/memory/file_memory.py
"""Uzun sureli hafiza: bagimliliksiz JSON dosyasi."""

from __future__ import annotations

import asyncio
import json
import secrets
import time
from dataclasses import asdict, dataclass

from ...config import config


@dataclass
class MemoryFact:
    id: str
    text: str
    created_at: float
    hits: int


class FileMemory:
    id = "file"

    def __init__(self) -> None:
        self._facts: list[MemoryFact] | None = None
        # Es zamanli yazmalar dosyayi bozmasin.
        self._lock = asyncio.Lock()

    async def list(self) -> list[MemoryFact]:
        if self._facts is not None:
            return self._facts
        try:
            raw = json.loads(config.memory_file.read_text("utf-8"))
            self._facts = [
                MemoryFact(
                    id=str(item.get("id", "")),
                    text=str(item.get("text", "")),
                    created_at=float(item.get("created_at", item.get("createdAt", 0))),
                    hits=int(item.get("hits", 1)),
                )
                for item in raw.get("facts", [])
                if item.get("text")
            ]
        except (OSError, json.JSONDecodeError, ValueError):
            self._facts = []
        return self._facts

    async def remember(self, text: str) -> MemoryFact | None:
        clean = " ".join(text.split())
        if not 3 <= len(clean) <= 240:
            return None

        facts = await self.list()
        normalized = clean.casefold()
        for fact in facts:
            if fact.text.casefold() == normalized:
                fact.hits += 1
                await self._persist()
                return fact

        fact = MemoryFact(
            id=f"m_{secrets.token_hex(5)}", text=clean, created_at=time.time(), hits=1
        )
        facts.append(fact)
        if len(facts) > config.memory_max_facts:
            # Sinir asilirsa en az kullanilan ve en eski kayitlar dusurulur.
            facts.sort(key=lambda f: (f.hits, f.created_at), reverse=True)
            del facts[config.memory_max_facts :]
        await self._persist()
        return fact

    async def forget(self, fact_id: str) -> None:
        facts = await self.list()
        self._facts = [fact for fact in facts if fact.id != fact_id]
        await self._persist()

    async def clear(self) -> None:
        self._facts = []
        await self._persist()

    async def _persist(self) -> None:
        async with self._lock:
            payload = {"version": 1, "facts": [asdict(fact) for fact in self._facts or []]}
            config.memory_file.parent.mkdir(parents=True, exist_ok=True)
            config.memory_file.write_text(
                json.dumps(payload, ensure_ascii=False, indent=2), "utf-8"
            )


memory = FileMemory()

File: server/src/providers/memory/fileMemory.ts
import fs from 'node:fs/promises';
import path from 'node:path';
import { config } from '../../config.js';
import type { MemoryFact, MemoryProvider } from '../types.js';

/** Uzun sureli hafiza: bagimliliksiz JSON dosyasi. */
export class FileMemoryProvider implements MemoryProvider {
  readonly id = 'file';
  private cache: MemoryFact[] | null = null;
  private writing: Promise<void> = Promise.resolve();

  async list(): Promise<MemoryFact[]> {
    if (this.cache) return this.cache;
    try {
      const raw = await fs.readFile(config.memory.file, 'utf8');
      const parsed = JSON.parse(raw) as { facts?: MemoryFact[] };
      this.cache = Array.isArray(parsed.facts) ? parsed.facts : [];
    } catch {
      this.cache = [];
    }
    return this.cache;
  }

  async remember(text: string): Promise<MemoryFact | null> {
    const clean = text.trim().replace(/\s+/g, ' ');
    if (clean.length < 3 || clean.length > 240) return null;

    const facts = await this.list();
    const normalized = clean.toLocaleLowerCase('tr');
    const existing = facts.find((f) => f.text.toLocaleLowerCase('tr') === normalized);
    if (existing) {
      existing.hits += 1;
      await this.persist();
      return existing;
    }

    const fact: MemoryFact = {
      id: `m_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 7)}`,
      text: clean,
      createdAt: Date.now(),
      hits: 1,
    };
    facts.push(fact);
    // Sinir asilirsa en az kullanilan ve en eski kayitlar dusurulur.
    if (facts.length > config.memory.maxFacts) {
      facts.sort((a, b) => b.hits - a.hits || b.createdAt - a.createdAt);
      facts.length = config.memory.maxFacts;
    }
    await this.persist();
    return fact;
  }

  async forget(id: string): Promise<void> {
    const facts = await this.list();
    const index = facts.findIndex((f) => f.id === id);
    if (index >= 0) {
      facts.splice(index, 1);
      await this.persist();
    }
  }

  async clear(): Promise<void> {
    this.cache = [];
    await this.persist();
  }

  private async persist(): Promise<void> {
    const snapshot = { version: 1, facts: this.cache ?? [] };
    // Yazmalari siraya al: es zamanli istekler dosyayi bozmasin.
    this.writing = this.writing.then(async () => {
      await fs.mkdir(path.dirname(config.memory.file), { recursive: true });
      await fs.writeFile(config.memory.file, JSON.stringify(snapshot, null, 2), 'utf8');
    });
    await this.writing;
  }
}

export const memory = new FileMemoryProvider();
